nomos-mix/core/src/message_blend/temporal.rs

```diff
@@ -94,33 +78,72 @@ where
         if let Some(timer) = self.release_timer.as_mut() {
             if timer.as_mut().poll(cx).is_ready() {
                 self.release_timer.take(); // Reset timer after it's done
-                if let Some(msg) = self.queue.pop_front() {
-                    // Release the 1st message in the queue if it exists.
-                    return Poll::Ready(Some(msg));
-                }
+                return Poll::Ready(Some(()));
             }
         }
+        Poll::Pending
+    }
+}
 
+/// [`TemporalProcessor`] delays messages randomly to hide timing correlation
+/// between incoming and outgoing messages from a node.
+///
+/// See the [`Stream`] implementation below for more details on how it works.
+pub struct TemporalProcessor<M, S> {
+    // All scheduled messages
+    queue: VecDeque<M>,
+    scheduler: S,
+}
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
+pub struct TemporalSchedulerSettings {
+    pub max_delay_seconds: u64,
+}
+
+impl<M, S> TemporalProcessor<M, S> {
+    pub(crate) fn new(scheduler: S) -> Self {
+        Self {
+            queue: VecDeque::new(),
+            scheduler,
+        }
+    }
+    /// Schedule a message to be released later.
+    pub(crate) fn push_message(&mut self, message: M) {
+        self.queue.push_back(message);
+    }
+}
+
+impl<M, S> Stream for TemporalProcessor<M, S>
+where
+    M: Unpin,
+    S: Stream<Item = ()> + Unpin,
+{
+    type Item = M;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.scheduler.poll_next_unpin(cx).is_ready() {
+            return Poll::Ready(self.queue.pop_front());
+        };
         Poll::Pending
     }
```

Comment:
I think we should return `Poll::Pending` instead of `Poll::Ready(None)` when the `self.queue.pop_front()` returns `None`.

Reply:
I fixed this 75e237e to merge this PR now and start implementing simulation.